### PR TITLE
Add a method argument

### PR DESCRIPTION
**Pull request by vindarel**
for #84 

My new tests are failing, this in to start a discussion and welcome any remarks if you have some.

### CODE REVIEW

**Comment by abingham** [comment]
This approach looks just fine. As you say, the test is failing, but it seems to be failing in the appropriate way. Once you update the canned test case, I think the functionality will be fine.

I have a few comments on the code, but I'll make those inline.

File: traad/app.py [hunk 1]
```diff
@@ -186,6 +186,18 @@ def remove_argument_view():
                                args['offset'])
 
 
+@app.post('/refactor/add_argument')
+def add_argument_view():
+    args = bottle.request.json
+    return standard_async_task(bottle.request.app.project.add_argument,
+                               args['index'],
```
Comment (abingham) [inline traad/app.py]: I think these are in the wrong order. `add_argument` seems to take path, offset, index, name, default, and value, in that order.

File: traad/rope/change_signature.py [hunk 2]
```diff
@@ -84,3 +84,33 @@ def remove_argument(self,
...
+        """Add an argument to a method.
+
+        Args:
+          self: The Project on which this operates.
+          state: The TaskState for this refactoring.
+          path: The path of the file/directory to query.
+          offset: The offset in the resource of the method signature.
+          index: The position of the new argument in the signature.
+          name: The name of the new argument.
+          default:
```
Comment (abingham) [inline traad/rope/change_signature.py]: We need documentation for `default` and `value`.

**Comment by vindarel** [comment]
I had another look on the tests and I don't understand how to choose the method to add an argument to. For example, to `projects.basic.a_free_func`. Likewise I didn't get what the `test_cross_remove_argument` really works. Will try again.

**Comment by abingham** [comment]
I've started the process of converting the test suite over to using pytest. It seems to be quite an improvement over the current version. I think it's a bit cleaner, so you might look at it for guidance.

In fact, if you want, you could add your new functionality on the master branch without tests. Then, we could add your tests to the pytest branch. I leave that up to you.

**Comment by abingham** [comment]
The pytest branch is complete now, but don't want to merge it in if it's going to derail or confuse your work. In that branch I've completely removed the unittest-based tests and am now using pytest universally.

If you don't mind, I can merge it in now and you can make your test using pytest. Or I can wait until you're done, then rebase the pytest stuff atop your changes and translate the new test to pytest. Which would you prefer?

**Comment by vindarel** [comment]
Please go ahead on merge pytest, it won't hurt me.

**Comment by vindarel** [comment]
Well, I thought I had rebased enough but that is not the case…

Anyway, I wanted to say I understand the tests with your explanations (the role of the offset), but it seems regular tests are failing:

```
tests/test_rename.py .

==================================================== FAILURES ====================================================
_______________________________________________ test_get_children ________________________________________________

activate_package = <function f at 0x7f9813dc91b8>, start_project = <function f at 0x7f9813dc9578>

    def test_get_children(activate_package, start_project):
        activate_package(package='basic', into='main')
        proj = start_project('main')
    
>       assert (sorted(proj.get_children('basic').get()) ==
                [('basic/__init__.py', False),
                 ('basic/bar.py', False),
                 ('basic/foo.py', False),
                 ('basic/overrides.py', False)])

tests/test_project.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.virtualenvs/traad/local/lib/python2.7/site-packages/pykka/threading.py:52: in get
    compat.reraise(*self._data['exc_info'])
../../.virtualenvs/traad/local/lib/python2.7/site-packages/pykka/compat.py:12: in reraise
    exec('raise tp, value, tb')
../../.virtualenvs/traad/local/lib/python2.7/site-packages/pykka/actor.py:201: in _actor_loop
    response = self._handle_receive(message)
../../.virtualenvs/traad/local/lib/python2.7/site-packages/pykka/actor.py:295: in _handle_receive
    return callee(*message['args'], **message['kwargs'])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Project("/home/vince/bacasable/traad/tests/active/main/"), path = 'basic'

    def get_children(self, path):
        '''Get a list of all child resources of a given path.
    
            ``path`` may be absolute or relative. If ``path`` is relative,
            then it must to be relative to the root of the project.
    
            Args:
              path: The path of the file/directory to query.
    
            Returns: A list of tuples of the form (path,
              is_folder).
    
            '''
    
>       path = self._to_relative_path(path)
E       AttributeError: 'Project' object has no attribute '_to_relative_path'

../../.virtualenvs/traad/local/lib/python2.7/site-packages/traad-0.11-py2.7.egg/traad/rope/project.py:171: AttributeError
=============================================== 1 tests deselected ===============================================
=============================== 1 failed, 24 passed, 1 deselected in 0.50 seconds ========================
```

**Comment by abingham** [comment]
@vindarel I've cherry-picked your changes cleanly onto master in [the add-method-argument branch](https://github.com/abingham/traad/tree/add-method-argument). Please have a look at that branch and let me know if accurately reflects that changes you've made. If so, then I'll merge it in and cancel this pull request.

Your work was good, but this cherry-pick was the simplest way for me to get a clean merge. I had to clean up a few things, but that was mostly a result of the confusing state of my code that you had to work with.

So let me know if you see anything in that branch that you don't agree with, and thanks again for your work on this. I know it wasn't particularly easy, but I really appreciate contributions like this!

**Comment by vindarel** [comment]
oh, so nice this cherry pick :) 
the changes look ok and the new test passes \o/ (but `test_rename` fails).

Now on emacs-traad ?

(noo it was very doable, and it was nice you letting me do )

**Comment by abingham** [comment]
Great! I'll sort this out right away.

And yes, it would be very nice if you could update emacs-traad, too. Thanks!

**Pull request closed without merging**